packages/react-reconciler/src/__tests__/ReactDoubleInvokeEvents-test.internal.js

```diff
@@ -0,0 +1,290 @@
+    function App({text}) {
+      React.useEffect(() => {
+        Scheduler.unstable_yieldValue('useEffect mount');
+        return () => Scheduler.unstable_yieldValue('useEffect unmount');
+      });
+
+      React.useLayoutEffect(() => {
+        Scheduler.unstable_yieldValue('useLayoutEffect mount');
+        return () => Scheduler.unstable_yieldValue('useLayoutEffect unmount');
+      });
+
+      return text;
+    }
+    ReactNoop.act(() => {
+      ReactNoop.render(<App text={'mount'} />);
+    });
+
+    if (__DEV__) {
+      expect(Scheduler).toHaveYielded([
+        'useLayoutEffect mount',
+        'useLayoutEffect unmount',
+        'useLayoutEffect mount',
+        'useEffect mount',
+        'useEffect unmount',
+        'useEffect mount',
+      ]);
+    } else {
+      expect(Scheduler).toHaveYielded([
+        'useLayoutEffect mount',
+        'useEffect mount',
+      ]);
+    }
+
+    ReactNoop.act(() => {
+      ReactNoop.render(<App text={'update'} />);
+    });
+
+    if (__DEV__) {
+      expect(Scheduler).toHaveYielded([
+        'useLayoutEffect unmount',
+        'useLayoutEffect mount',
+        'useLayoutEffect unmount',
+        'useLayoutEffect mount',
+        'useEffect unmount',
+        'useEffect mount',
+        'useEffect unmount',
+        'useEffect mount',
+      ]);
+    } else {
+      expect(Scheduler).toHaveYielded([
+        'useLayoutEffect unmount',
+        'useLayoutEffect mount',
+        'useEffect unmount',
+        'useEffect mount',
+      ]);
+    }
+  });
+  it('first useEffect double invokes before second useEffect is called', () => {
+    function App({text}) {
+      React.useEffect(() => {
+        Scheduler.unstable_yieldValue('useEffect One mount');
+        return () => Scheduler.unstable_yieldValue('useEffect One unmount');
+      });
+
+      React.useEffect(() => {
+        Scheduler.unstable_yieldValue('useEffect Two mount');
+        return () => Scheduler.unstable_yieldValue('useEffect Two unmount');
+      });
+
+      return text;
+    }
+
+    ReactNoop.act(() => {
+      ReactNoop.render(<App text={'mount'} />);
+    });
+
+    if (__DEV__) {
+      expect(Scheduler).toHaveYielded([
+        'useEffect One mount',
+        'useEffect One unmount',
+        'useEffect One mount',
+        'useEffect Two mount',
+        'useEffect Two unmount',
+        'useEffect Two mount',
+      ]);
+    } else {
+      expect(Scheduler).toHaveYielded([
+        'useEffect One mount',
+        'useEffect Two mount',
+      ]);
+    }
+
+    ReactNoop.act(() => {
+      ReactNoop.render(<App text={'update'} />);
+    });
+
+    if (__DEV__) {
+      expect(Scheduler).toHaveYielded([
+        'useEffect One unmount',
+        'useEffect Two unmount',
+        'useEffect One mount',
+        'useEffect One unmount',
+        'useEffect One mount',
+        'useEffect Two mount',
+        'useEffect Two unmount',
+        'useEffect Two mount',
```

Comment:
I think this is actually a regression.

We guarantee all destroy functions run before any mount functions. This is important because interleaving destroy/create effects between sibling components might cause components to interfere with each other (e.g. a destroy function in one component may unintentionally override a ref value set by a create function in another component). This was something we recently fixed, via #17945.

The approach implemented in this PR seems to regress, since in DEV mode, when we double-invoke things, we break the order (as shown in this unit test).

I think we should rethink the approach here. Let's talk?

Reply:
Talked to Sebastian last night after submitting this diff and we decided to only do this in mount (because unmount is called already during update). I can change the approach to not interleave the destroy/create function as well. That mirrors the hidden/suspense case better anyway!

Reply:
Ah! Yeah, I think that makes sense.

Cool. Ping me again when this is ready for review (or we can talk about it or pair on it if you'd like too)

Reply:
Yea the interleaving thing is important, but also consider passive effects.

It should be:

1) all mutation effects
2) all layout effects (including didmount)
3) all passive effects
4) all layout effect destroys (including willunmount)
5) all passive effect destroys
6) all layout effects (including didmount) again
7) all passive effects again

I think you'll need to take the approach of running the whole traversals of the tree over and over again.

Reply:
The thing we're simulating is effects three commits: 1) initial commit 2) commit to hide 3) commit to show.

So you basically have to replicate everything in commitRoot (but not exactly).